### DISCUSSION TITLE
How to determine if PublishAsync (not Jetstream's) succeeded?

### DISCUSSION THREAD

**Discussion by robertmircea** [nats-io/nats.net#278] [Q&A] [answered]
Is there a way to check if `NatsConnection`'s PublishAsync method has succeeded?

 `public ValueTask PublishAsync(string subject, NatsHeaders? headers = default, string? replyTo = default, NatsPubOpts? opts = default, CancellationToken cancellationToken = default)`

For `Jetstream`'s PublishAsync, I can check `PubAckResponse`.

**Comment by mtmk** [comment, marked as answer]
There are no delivery guaranties for core NATS publish. It's essentially fire-and-forget: https://docs.nats.io/nats-concepts/what-is-nats#nats-quality-of-service-qos.

However there are a couple of things you can do with `PublishAsync` to make sure at least the message is written to TCP stream as close as possible by setting the option `WaitUntilSent` to `true`. If `WaitUntilSent` is set to `false` messages will be queued up for faster send rate so another thing is to make sure to handle errors (usually serialization errors) by subscribing to the publish options `ErrorHandler`.
